src/logging.cc

```diff
@@ -2563,4 +2564,20 @@ void DisableLogCleaner() {
   log_cleaner.Disable();
 }
 
+long int LogMessage::CalcGmtOffset() {
```

Comment:
It feels like the method is out of place here and actually a better fit for `LogMessageTime`. It is also not necessary to recompute the GMT offset but rather compute it once (e.g., on demand) and cache the result.

Make also sure to follow the naming scheme of members used in the class for public methods and the source code formatting (missing spaces between `if/else` and braces) .

Reply:
Hi @sergiud , 

I could add the method to "LogMessageTime" but not sure how to add a unit testcase. 

"LogMessageTime" actually uses time info. from "LogMessage" class but it's not a member of this class. So I can't simply instantiate "LogMessage" & test the methods of "LogMessageTime"

Any efficient way you can think of to add a unittest here?  

Thanks, 
Vijay

Reply:
I believe there are two options:
1.  Test the GMT offset within the supplied custom prefix callback
2.  Make`LogMessageTime` available even if custom prefix functionality is not enabled and provide  `LogMessageTime` as a member variable (and corresponding getter) in `LogMessage`which seems more reasonable anyway but requires more refactoring.

Reply:
Hi @sergiud , 

1.  Tried adding test(`EXPECT_TRUE/FALSE`) in the Custom prefix callback(`PrefixAttacher`); works fine if the test succeeds. But if the test fails, somehow we are hitting infinite loop and it doesn't even print if the test is failing.  It exits with "Segmentation Fault"
Please have a look at following stack trace I got from GDB.

> (gdb) bt
#0  PrefixAttacher (s=..., l=..., data=0x6195c486) at /mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc:184
#1  0x00005555555f57a2 in google::LogMessage::Init (this=0x7fffffffd830,
    file=0x555555658590 "/mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc", line=337, severity=3,
    send_method=(void (google::LogMessage::*)(google::LogMessage * const)) 0x5555555f5e56 <google::LogMessage::SendToLog()>)
    at /mnt/d/Learning/git/Google/glog/src/logging.cc:1651
#2  0x00005555555f4e7b in google::LogMessage::LogMessage (this=0x7fffffffd830,
    file=0x555555658590 "/mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc", line=337, severity=3)
    at /mnt/d/Learning/git/Google/glog/src/logging.cc:1538
#3  0x00005555555f980b in google::LogMessageFatal::LogMessageFatal (this=0x7fffffffd830,
    file=0x555555658590 "/mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc", line=337)
    at /mnt/d/Learning/git/Google/glog/src/logging.cc:2477
#4  0x00005555555db397 in NoAllocNewHook () at /mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc:337
#5  0x00005555555d89ab in operator new (size=392) at /mnt/d/Learning/git/Google/glog/src/googletest.h:629
#6  0x000055555560acad in testing::Message::Message() ()
#7  0x00005555555d9b98 in PrefixAttacher (s=..., l=..., data=0x7fffffffe150) at /mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc:208
#8  0x00005555555f57a2 in google::LogMessage::Init (this=0x7fffffffdc10,
    file=0x555555658590 "/mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc", line=337, severity=3,
    send_method=(void (google::LogMessage::*)(google::LogMessage * const)) 0x5555555f5e56 <google::LogMessage::SendToLog()>)
    at /mnt/d/Learning/git/Google/glog/src/logging.cc:1651
#9  0x00005555555f4e7b in google::LogMessage::LogMessage (this=0x7fffffffdc10,
    file=0x555555658590 "/mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc", line=337, severity=3)
    at /mnt/d/Learning/git/Google/glog/src/logging.cc:1538
#10 0x00005555555f980b in google::LogMessageFatal::LogMessageFatal (this=0x7fffffffdc10,
    file=0x555555658590 "/mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc", line=337)
    at /mnt/d/Learning/git/Google/glog/src/logging.cc:2477
#11 0x00005555555db397 in NoAllocNewHook () at /mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc:337
#12 0x00005555555d89ab in operator new (size=4) at /mnt/d/Learning/git/Google/glog/src/googletest.h:629
#13 0x00005555555db43d in DeathNoAllocNewHook_logging_Test::TestBody (this=0x5555556bc520)
    at /mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc:352
#14 0x000055555564382e in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
#15 0x000055555563ba07 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
#16 0x0000555555610524 in testing::Test::Run() ()
#17 0x0000555555610f66 in testing::TestInfo::Run() ()
#18 0x000055555561189b in testing::TestSuite::Run() ()
#19 0x000055555562133b in testing::internal::UnitTestImpl::RunAllTests() ()
#20 0x0000555555644d5c in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
#21 0x000055555563cc45 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
#22 0x000055555561fa68 in testing::UnitTest::Run() ()
#23 0x00005555555e8e78 in RUN_ALL_TESTS () at /usr/local/include/gtest/gtest.h:2492
#24 0x00005555555d9fee in main (argc=1, argv=0x7fffffffe2a8) at /mnt/d/Learning/git/Google/glog/src/logging_custom_prefix_unittest.cc:248
(gdb)


2. We just need time info. generated by `LogMessage` which can be used directly to instantiate `LogMessageTime` object. The time info. is stored in `LogMessage` as a private data memeber (`LogMessageData* data_`). Instead of having `LogMessageTime` object as a member in the class  `LogMessage`, having a public method to get the time info from `LogMessage` seems efficient. 
Something like this, 
`void LogMessage::GetTimeInfo(struct::tm& time_struct, time_t& ts, int32_t& usecs)`

Please comment.

Thanks, 
Vijay

Reply:
1. You should move the test out of the callback and, for instance, write the result into the user supplied parameter.
2. Sounds reasonable for a private method. I would still cache the result to avoid multiple time info queries per single log message.

Reply:
will calculate the "gmtoffset" in the constructor of the `LogMessageTime` and store it as a data member. Should be fine, right?

Reply:
I believe so, yes.